### PR TITLE
Compress events using fflate (gzip-js)

### PR DESCRIPTION
**Pull request by macobo**
## Changes

After this we send gzipped data to the /e/ endpoint (if supported) - this is much better than lzstring and faster!

TODO: 
- [x] gzip tests are unstable
- [x] content-length seems 50% too large - might be an encoding issue
- [x] posthog-side PR - https://github.com/PostHog/posthog/pull/2701
- [x] drop lzstring (will be a separate PR)

## Checklist
- [x] Tests for new code (if applicable)
- [x] TypeScript definitions (module.d.ts) updated and in sync with library exports (if applicable)

### CODE REVIEW

File: src/posthog-core.js [hunk 1]
```diff
@@ -339,9 +340,17 @@ PostHogLib.prototype._handle_queued_event = function (url, data, options) {
...
-    if (this.get_config('_capture_metrics')) {
-        this._send_request(url, jsonData, { ...options, plainJSON: true }, callback)
-    } else if (this.compression['lz64'] || (options.compression && options.compression === 'lz64')) {
+    if (this.compression['gzip-js']) {
+        // :TRICKY: This is encoded as latin-1/iso-8859-1, needs to be decoded accordingly on the server.
+        const compressed = strFromU8(gzipSync(strToU8(jsonData)), true)
+
+        this._send_request(
+            url,
+            compressed,
```
Comment (paolodamico) [inline src/posthog-core.js:349]: Curious on the rationale for passing data this way instead of the same way we do it for LZ64 compression (`{ data: LZString.compressToBase64(jsonData), compression: 'lz64' }`), if this is because the data is binary, then we could have some unhandled edge cases (e.g. if `config.test = True`, see https://github.com/PostHog/posthog-js/blob/20ee66b81ae0902ffbe3bf17dad3a4a502266076/src/posthog-core.js#L384)

**Review by paolodamico — COMMENTED** [review]
Overall looking pretty good. The `fflate` benchmarks and metrics look pretty amazing.

File: src/posthog-core.js (hunk 1 above)
Comment (macobo) [reply src/posthog-core.js:349]: Great comments!

The lines you mentioned lines are (effectively) dead, but require further refactoring to make it so in practice. I will do so soon though.

The reason we're passing data this way has to do with the request itself:
1. We can do a Content-Type: application/x-www-form-urlencoded request (as with lzstring) - this requires base64-encoding the data, which ends up wasting bytes
2. Or we can do a Content-Type: text/plain (as we do here) request - this doesn't require base64-encoding the data if we can get both of the ends to encode/decode the data correctly.

Normally 1. is simpler but takes more space - however I couldn't get base64-encoded gzip working without doing significant js-side changes so decided to instead extend the posthog server.

File: src/__tests__/posthog-core.js [hunk 2]
```diff
@@ -261,3 +261,40 @@ describe('_handle_unload()', () => {
         })
     })
 })
+
+describe('__compress_and_send_json_request', () => {
+    given('subject', () => () =>
+        given.lib.__compress_and_send_json_request('/e/', given.jsonData, given.options, jest.fn())
+    )
+
+    given('jsonData', () => JSON.stringify({ large_key: 'a' * 300 }))
```
[the remark this answers is not in the document]
Comment (macobo) [reply src/__tests__/posthog-core.js:270]: Yup - these tests are failing right now - will iterate tomorrow.

File: src/posthog-core.js (hunk 1 above)
Comment (paolodamico) [reply src/posthog-core.js:349]: That makes total sense! Seems like option 2 might be best in the long run

File: src/__tests__/posthog-core.js [hunk 3]
```diff
@@ -261,3 +261,40 @@ describe('_handle_unload()', () => {
...
+        compression: given.compression,
+        _send_request: jest.fn(),
+    }))
+
+    it('handles base64 compression', () => {
+        given('compression', () => ({}))
+
+        given.subject()
+
+        expect(given.overrides._send_request.mock.calls).toMatchSnapshot()
```
[the remark this answers is not in the document]
Comment (macobo) [reply src/__tests__/posthog-core.js:282]: I changed this logic - now the file will be visible. :)

File: src/__tests__/posthog-core.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (macobo) [reply src/__tests__/posthog-core.js]: Fixed

**Review by paolodamico — APPROVED** [review]
lgtm!
